invoice/src/main/java/org/killbill/billing/invoice/usage/ContiguousIntervalCapacityUsageInArrear.java

```diff
@@ -0,0 +1,138 @@
+import org.killbill.billing.catalog.api.Usage;
+import org.killbill.billing.invoice.api.InvoiceItem;
+import org.killbill.billing.invoice.model.UsageInvoiceItem;
+import org.killbill.billing.invoice.usage.details.UsageCapacityInArrearDetail;
+import org.killbill.billing.invoice.usage.details.UsageConsumableInArrearTierUnitDetail;
+import org.killbill.billing.invoice.usage.details.UsageInArrearDetail;
+import org.killbill.billing.invoice.usage.details.UsageInArrearTierUnitDetail;
+import org.killbill.billing.usage.RawUsage;
+import org.killbill.billing.usage.api.RolledUpUnit;
+import org.killbill.billing.util.config.definition.InvoiceConfig.UsageDetailMode;
+import org.killbill.billing.util.jackson.ObjectMapper;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Joiner;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Lists;
+
+import static org.killbill.billing.invoice.usage.UsageUtils.getCapacityInArrearTier;
+
+public class ContiguousIntervalCapacityUsageInArrear extends ContiguousIntervalUsageInArrear {
+
+    private static final Logger log = LoggerFactory.getLogger(ContiguousIntervalCapacityUsageInArrear.class);
+
+    private static final ObjectMapper objectMapper = new ObjectMapper();
+
+    public ContiguousIntervalCapacityUsageInArrear(final Usage usage,
+                                                   final UUID accountId,
+                                                   final UUID invoiceId,
+                                                   final List<RawUsage> rawSubscriptionUsage,
+                                                   final LocalDate targetDate,
+                                                   final LocalDate rawUsageStartDate,
+                                                   final UsageDetailMode usageDetailMode,
+                                                   final InternalTenantContext internalTenantContext) {
+        super(usage, accountId, invoiceId, rawSubscriptionUsage, targetDate, rawUsageStartDate, usageDetailMode, internalTenantContext);
+    }
+
+
+
+    @Override
+    protected void populateResults(final LocalDate startDate, final LocalDate endDate, final Iterable<InvoiceItem> billedItems, final BigDecimal billedUsage, final BigDecimal toBeBilledUsage, final UsageInArrearDetail toBeBilledUsageDetails, final boolean areAllBilledItemsWithDetails, final List<InvoiceItem> result) {
+        // Compute final amount by subtracting  amount that was already billed.
+        if (!billedItems.iterator().hasNext() || billedUsage.compareTo(toBeBilledUsage) < 0) {
+            final BigDecimal amountToBill = toBeBilledUsage.subtract(billedUsage);
+
+            if (amountToBill.compareTo(BigDecimal.ZERO) > 0) {
+                    final String itemDetails = areAllBilledItemsWithDetails ? toBeBilledUsageDetails.toJson(objectMapper) : null;
```

Comment:
Why do we need to check `areAllBilledItemsWithDetails`? `toBeBilledUsageDetails` is independent of `areAllBilledItemsWithDetails` for capacity.

Reply:
For now it is, but i think the discussion is open; we have to decide exactly which info we put in this json: If the indo we add is independent of any previously billed amount, -- which it is with my current implementation, what you are saying makes sense.  Now i went into the trouble of splitting the json between the 2 modes, with the intention to revisit this specifically. So will keep it the way it is for now, but more thoughts is needed here.

Reply:
So more context, current json for capacity is the following -- see new test i wrote with existing data usage for past period `TestContiguousIntervalCapacityInArrear#testMultipleItemsAndTiersWithExistingItems`:

Total usage amount is `90.00` -- taking into account `10.00` that was already billed. Now new detail json looks like:

```
{
  "tierDetails": [
    {
      "tier": 2,
      "tierUnit": "BAR",
      "tierPrice": 10,
      "quantity": 200
    },
    {
      "tier": 3,
      "tierUnit": "FOO",
      "tierPrice": 100,
      "quantity": 60
    }
  ],
  "amount": 100
}
```

It currently reflects the total number of units in this period, but i think we should add another field -- this is mostly why i broke the json to be different between capacity and consumable -- to include the part that was already paid. So maybe an extra field -- at the top level -- like `previouslyBilledAmount=10.00`